stdlib/src/collections/list.mojo

```diff
@@ -568,3 +568,46 @@ struct List[T: CollectionElement](CollectionElement, Sized, Boolable):
         """
         var ref = Reference(self)
         return _ListIter[T, mutability, self_life, False](len(ref[]), ref)
+
+# TODO: Convert functions below to list methods when issue 1876 is resolved.
+@always_inline
+fn index[C: ComparableCollectionElement](self: List[C], owned value: C, start: Int = 0) raises -> Int:
```

Comment:
If I may suggest, making this a `@staticmethod` instead of a free function. A few reasons I can think of:

* it allows a user to call this method without importing anything.
* It allows `__type_of(my_list).index(my_list, "hello")` which is easy to search and replace when #1876 is fixed
* When #1876 is fixed, users and us stdlib authors won't have to change the imports
* When #1876 is fixed, we won't have to move the code into the struct, avoiding a big diff due to indentation (for free functions, the indentation is 0 spaces, for methods it's 4 spaces)
* For built-in structs, users don't import anything to use them. So it's not obvious where they should import the free function from: `from ??? import index`. For us, standard library devs, it's obvious what the import is, but for Mojo users, they'll have to look at the docs to know the correct import.
* In this PR, the `index` method is imported in `stdlib/src/collections/__init__.mojo`, what do we do when multiple collections have a method with the same name, like `__eq__` ?
* For methods common to multiple types, it's going to get confusing. Let's say the user needs `__eq__` on both a List and a Dict in the same file, then users will have to do 
```mojo
from collections.list import __eq__
from collections.dict import __eq__
```
and it's going to get confusing.

For all those reasons, I suggest going for the syntax `__type_of(my_list).index(my_list, "hello")` which will lead to less issues in the future.

You can take a look at the way to make this syntax happen in this PR: https://github.com/modularml/mojo/pull/2323